apps/block_scout_web/assets/js/omni.js

```diff
@@ -0,0 +1,163 @@
+  ${JSON.stringify(details, null, 4)}
+  </pre>`;
+  return div;
+};
+
+// expected return {txHash1:true, txHas2:true...}
+const checkTx = async (elements) => {
+  const response = await fetch(url + "/checkOmniTxHashes", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      txHashes: elements,
+    }),
+  });
+  if (response.status == 400) {
+    await response.json().then(console.error);
+    return undefined;
+  }
+  return response.json();
+};
+
+const getTxDetails = async (omniTxHash) => {
+  // dummy response
+  const response = await fetch(url + `/omniTx/${omniTxHash}`);
+  const data = await response.json();
+  if (data.length === 0) {
+    return undefined;
+  }
+  return data;
+};
+
+const getAllTxHashElements = () => {
+  return document.querySelectorAll('[data-test="transaction_hash_link"]');
+};
+
+const handletxHashLinks = async () => {
+  const elements = getAllTxHashElements();
+  if (elements.length == 0) {
+    return;
+  }
+  // check local cache
+  const unknownTxList = [];
+  for (let i of elements) {
+    if (i.parentNode.childElementCount <= 2 && txMap[i.innerHTML] === true) {
+      console.log("from cache " + i.innerHTML);
+      i.parentNode.appendChild(customMarker());
+    } else if (txMap[i.innerHTML] === undefined) {
+      unknownTxList.push(i.innerHTML);
+    }
+  }
+  if (unknownTxList.length == 0) {
+    return;
+  }
+  const xChainTxs = await checkTx(unknownTxList);
+  if (xChainTxs === undefined) {
+    return;
+  }
+  for (let i of elements) {
+    if (xChainTxs[i.innerHTML] === true) {
+      txMap[i.innerHTML] = true;
+      console.log("from api " + i.innerHTML);
+      i.parentNode.appendChild(customMarker());
+    } else if (txMap[i.innerHTML] == undefined) {
+      txMap[i.innerHTML] = false;
+    }
+  }
+};
+
+const displayTxDetails = async (txHash) => {
+  const elements = document.querySelectorAll(
+    '[class="card js-ad-dependant-mb-3"]'
+  );
+  if (elements.length != 1) {
+    console.log("could not find unique tx view element");
+    return;
+  }
+
+  const details = await getTxDetails(txHash);
+  if (details === undefined) return;
+
+  // if tx has cross chain data display
+  elements[0].appendChild(customXChainDataView(details));
+};
+
+// sets next repeat only after the function call completes
+const repeatLoop = async (timeMS) => {
+  try {
+    await handletxHashLinks();
+  } catch (err) {
+    console.log(err);
+  } finally {
+    setTimeout(repeatLoop, timeMS);
+  }
+};
+
+const watch = async () => {
+
+  // update api url depending on blockscouts deployed environment
+  if (window.location.hostname.startsWith("testnet")) {
+    url = "testnet-1-xchain.explorer.omni.network";
+  } else if (window.location.hostname.startsWith("staging")) {
+    url = "staging-xchain.explorer.omni.network";
+  }
+
+  // handle for homepage
+  if (window.location.pathname === "/") {
+    repeatLoop(1500);
```

Comment:
is repeatLoop mechanism how blockscout ui is doing api calls, or this is custom thing for us? is there a difference in rate in which the rest of blockscout ui is polling for data

Reply:
Blockscout is server side rendered mostly and uses websockets for pushing data from backend to frontend. Our changes are patch work changes made to add extra info in the UI so it's compeltely separate from blockscout.

Reply:
okay with this for now. some thoughts.

- can use setInterval instead of setTimeout if you'd like. also, it shouldn't be an issue because we only start a loop at page load, but in case that changes, we should clear our timeouts / intervals when they're no longer needed. to avoid running loops in the background that we don't need
- rather than using setInterval / setTimeout, we might consider a [mutation observer](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), to watch for changes in say, a list of transaction. and then react to new elements accordingly. should make our reaction to new txs a bit smoother
- at some point would be nice to inject our omni xchain details into the websocket new_transactions channels, to avoid a delay in displaying xchain markers

Reply:
1. I use setTimeout because I use the pattern where we run the next iteration wait only when the previous one completes, setInterval just fires automatically without waiting for the actual function to finish as far as I have seen.

2. Not using mutation observer because if you see our testnet env the incoming tx rate is very high we don't want to make an api call for each and every single new incoming tx. The interval approach batches a few txs and reduces the load on the servers significantly as the expense of a small delay in the UI which I think is probably fine. We are not going to settle on this patch work ui and it'll probably never be the solution. We will have to build a new UI from scratch using something like react leveraging the blockscout apis and xchain indexer apis in more optimized way.